sdk/search/Azure.Search/src/Models/SearchResult{T}.cs

```diff
@@ -0,0 +1,166 @@
+
+        /// <summary>
+        /// The document found by the search query.
+        /// </summary>
+        public T Document { get; internal set; }
+
+        /// <summary>
+        /// Initializes a new instance of the SearchResult class.
+        /// </summary>
+        internal SearchResult() { }
+    }
+
+    /// <summary>
+    /// JsonConverterFactory to create closed instances of
+    /// <see cref="SuggestResultsConverter{T}"/>.
+    /// </summary>
+    internal class SearchResultConverterFactory : ModelConverterFactory
+    {
+        protected override Type GenericType => typeof(SearchResult<>);
+        protected override Type GenericConverterType => typeof(SuggestResultConverter<>);
+    }
+
+    /// <summary>
+    /// Convert from JSON to <see cref="SearchResult{T}"/>.
+    /// </summary>
+    /// <typeparam name="T">
+    /// The .NET type that maps to the index schema. Instances of this type can
+    /// be retrieved as documents from the index.
+    /// </typeparam>
+    internal class SuggestResultConverter<T> : JsonConverter<SearchResult<T>>
+    {
+        /// <summary>
+        /// Serializing SearchResult isn't supported as it's an output only
+        /// model type.  This always fails.
+        /// </summary>
+        /// <param name="writer">The JSON writer.</param>
+        /// <param name="value">The search result.</param>
+        /// <param name="options">Serialization options.</param>
+        public override void Write(Utf8JsonWriter writer, SearchResult<T> value, JsonSerializerOptions options) =>
+            throw new NotSupportedException($"{nameof(SearchResult<T>)} cannot be serialized to JSON.");
+
+        /// <summary>
+        /// Parse a SearchResult and its model.
+        /// </summary>
+        /// <param name="reader">The JSON reader.</param>
+        /// <param name="typeToConvert">The type to parse into.</param>
+        /// <param name="options">Serialization options.</param>
+        /// <returns>The deserialized search result.</returns>
+        public override SearchResult<T> Read(
+            ref Utf8JsonReader reader,
+            Type typeToConvert,
+            JsonSerializerOptions options)
+        {
+            Debug.Assert(options != null);
+            SearchResult<T> result = new SearchResult<T>();
+
+            // Clone the reader so we can get the search text property without
+            // advancing the reader over any properties needed to deserialize
+            // the user's model type.
+            Utf8JsonReader clone = reader;
+
+            // Keep track of how many of properties we've found so we can stop
+            // reading through the cloned object
+            int propertiesNeeded = 2;
+
+            clone.Expects(JsonTokenType.StartObject);
+            while (propertiesNeeded > 0 && clone.Read())
+            {
+                if (clone.TokenType == JsonTokenType.EndObject) { break; }
```

Comment:
Not throw? If 2 properties are expected, seems this would be exceptional.

Reply:
I suppose `Needed` is implying more than it should here.  There are a max of two properties we might read, but they might not always be present.  I don't think the service sends highlights if you didn't ask for them, for example.  We can stop after we've seen both and that's what it's tracking.

Reply:
Changed to use better names (and bools, per below)